### ISSUE TITLE
Fluid/Gas System

### ISSUE THREAD

**Issue by IntegratedQuantum** [PixelGuys/Cubyz#62]
Fluids and gases should be stored and simulated outside the normal voxel volume.

## Goals
- heavy liquid should flow downwards
- light gases should flow upwards
- biomes should have an associated liquid/fog block. Whenever a block in that biome is broken, the gap must get filled with the right fog/fluid block
- Fluids/Gases should exist in seperate memory to allow for underwater fences and similar
- Fluids should not propagate endlessly, flooding the entire world

## Relaxations
- fluid volume does not need to be preserved
- the flow does not need to be realistic
- big bodies of water like the ocean can be assumed as infinite, as long as they don't flood the entire underground

**Comment by IntegratedQuantum** [comment]
We cannot mix gases like on the left, since this will create an exponential amount of gas types.

**Comment by Argmaster** [comment]
Oxygen not included has a gas system which I believe allows only one gas type per tile, what we could have would be similarly constrained I think. Are there any other games with similar systems?

**Comment by IntegratedQuantum** [comment]
Mixed gases are also problematic since every border between different gas color/density needs a transparent face, which makes every biome with mixed gases a lag biome.

**Comment by archbirdplus** [comment]
Then we could have fluid glob together at a uniform density to reduce the number of faces to render. If they glob independently then we only have to deal with O(liquid dimensions) faces. We could also assign several levels of fluid to the same transparency to have the same dynamics without adding extra faces.

**Comment by IntegratedQuantum** [comment]
another approach was mentioned in #2388

**Comment by tillpp** [comment]
is there a concrete idea now you guys have settled in?
otherwise i would recommend my beloved hierarchical node approach here, even though it is more complex.

**Comment by IntegratedQuantum** [comment]
No, it's all open until someone puts in the effort to see how viable a given approach is (which is why this issue has the experimental tag).

**Comment by OneAvargeCoder193** [comment]
cubyz fluid system should be a bunch of chess games played by bots stacked on top of each other where white is water and black is lava or smth

**Comment by pgattic** [comment]
I found a cool article about fluid simulation. Don't know where we store articles but I thought I'd share it here in case it's noteworthy https://lisyarus.github.io/blog/posts/simulating-water-over-terrain.html

**Comment by TyrrelMeter** [comment]
Will gasses/liquids be able to have varying density?

**Comment by IntegratedQuantum** [comment]
liquids have varying density
